community/lucene-index/src/test/java/org/neo4j/kernel/impl/api/constraints/ConstraintRecoveryIT.java

```diff
@@ -0,0 +1,140 @@
+import org.neo4j.helpers.collection.Iterables;
+import org.neo4j.kernel.GraphDatabaseAPI;
+import org.neo4j.kernel.impl.api.index.IndexingService;
+import org.neo4j.kernel.impl.nioneo.xa.NeoStoreProvider;
+import org.neo4j.kernel.monitoring.Monitors;
+import org.neo4j.test.EphemeralFileSystemRule;
+import org.neo4j.test.TestGraphDatabaseFactory;
+import org.neo4j.test.impl.EphemeralFileSystemAbstraction;
+import org.neo4j.tooling.GlobalGraphOperations;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class ConstraintRecoveryIT
+{
+    private static final Label LABEL = DynamicLabel.label( "label1" );
+    @Rule
+    public EphemeralFileSystemRule fileSystemRule = new EphemeralFileSystemRule();
+    private GraphDatabaseAPI db;
+
+    @Test
+    public void shouldNotHaveAnIndexIfUniqueConstraintCreationOnRecoveryFails() throws IOException
+    {
+        // given
+        final EphemeralFileSystemAbstraction fs = fileSystemRule.get();
+        fs.mkdir( new File("/tmp") );
+        String pathToDb = "/tmp/bar2";
+
+        TestGraphDatabaseFactory dbFactory = new TestGraphDatabaseFactory();
+        dbFactory.setFileSystem( fs );
+
+        final EphemeralFileSystemAbstraction[] storeInNeedOfRecovery = new EphemeralFileSystemAbstraction[1];
+        final AtomicBoolean monitorCalled = new AtomicBoolean( false );
+
+        Monitors monitors = new Monitors();
+        monitors.addMonitorListener( new IndexingService.MonitorAdapter()
+        {
+            @Override
+            public void verifyDeferredConstraints()
+            {
+                monitorCalled.set( true );
+                db.getDependencyResolver().resolveDependency( NeoStoreProvider.class ).evaluate().getSchemaStore().flushAll();
+                storeInNeedOfRecovery[0] = fs.snapshot();
+            }
+        } );
+        dbFactory.setMonitors( monitors );
+
+
+        db = (GraphDatabaseAPI) dbFactory.newImpermanentDatabase( pathToDb );
+
+        try ( Transaction tx = db.beginTx() )
+        {
+            for ( int i = 0; i < 2; i++ )
+            {
+                Node node1 = db.createNode( LABEL );
+                node1.setProperty( "prop", true );
+            }
+
+            tx.success();
+        }
+
+        try ( Transaction tx = db.beginTx() )
+        {
+            db.schema().constraintFor( LABEL ).assertPropertyIsUnique( "prop" ).create();
+            fail("Should have failed with ConstraintViolationException");
+            tx.success();
+        }
+        catch ( ConstraintViolationException ignored )  { }
+
+        db.shutdown();
+
+        // when
+        dbFactory = new TestGraphDatabaseFactory();
+        dbFactory.setFileSystem( storeInNeedOfRecovery[0] );
+
+
+        System.out.println("starting recovery..");
+        db = (GraphDatabaseAPI) dbFactory.newImpermanentDatabase( pathToDb );
+
+        // then
+        assertTrue( monitorCalled.get() );
```

Comment:
Were you meant to clear `monitorCalled` before the second startup? It looks to me like it will be set to `true` the first startup already and just set to `true` again the second time.